homie-controller/src/lib.rs

```diff
@@ -182,67 +208,107 @@ impl HomieController {
         // our Arc to point to that, so that it is now a unique reference.
-        let devices = &mut *self.devices.lock().await;
+        let devices = &mut *self.devices.lock().unwrap();
         let devices = Arc::make_mut(devices);
 
+        // Collect MQTT topics to which we need to subscribe or unsubscribe here, so that the
+        // subscription can happen after the devices lock has been released.
+        let mut topics_to_subscribe: Vec<String> = vec![];
+        let mut topics_to_unsubscribe: Vec<String> = vec![];
+
         let parts = subtopic.split("/").collect::<Vec<&str>>();
-        match parts.as_slice() {
+        let event = match parts.as_slice() {
             [device_id, "$homie"] => {
                 if !devices.contains_key(*device_id) {
-                    self.new_device(devices, device_id, payload).await?;
-                    return Ok(Some(Event::DeviceUpdated {
-                        device_id: device_id.to_string(),
+                    log::trace!("Homie device '{}' version '{}'", device_id, payload);
+                    devices.insert((*device_id).to_owned(), Device::new(device_id, payload));
+                    let topic = format!("{}/{}/+", self.base_topic, device_id);
+                    topics_to_subscribe.push(topic);
+                    Some(Event::DeviceUpdated {
+                        device_id: (*device_id).to_owned(),
                         has_required_attributes: false,
-                    }));
+                    })
+                } else {
+                    None
                 }
             }
             [device_id, "$name"] => {
                 let device = get_mut_device_for(devices, "Got name for", device_id)?;
                 device.name = Some(payload.to_owned());
-                return Ok(Some(Event::device_updated(device)));
+                Some(Event::device_updated(device))
             }
             [device_id, "$state"] => {
                 let state = payload.parse()?;
                 let device = get_mut_device_for(devices, "Got state for", device_id)?;
                 device.state = state;
-                return Ok(Some(Event::device_updated(device)));
+                Some(Event::device_updated(device))
             }
             [device_id, "$implementation"] => {
                 let device = get_mut_device_for(devices, "Got implementation for", device_id)?;
                 device.implementation = Some(payload.to_owned());
-                return Ok(Some(Event::device_updated(device)));
+                Some(Event::device_updated(device))
             }
             [device_id, "$nodes"] => {
                 let nodes: Vec<_> = payload.split(",").collect();
                 let device = get_mut_device_for(devices, "Got nodes for", device_id)?;
+
                 // Remove nodes which aren't in the new list.
-                device.nodes.retain(|k, _| nodes.contains(&k.as_ref()));
+                device.nodes.retain(|node_id, node| {
```

Comment:
Ah. The pattern I was searching for was react's virtual dom.

If you have a device tree that's updated by these events, and a subscription tree that's managed by a separate process [the subscription tree might be a flat hashmap, but I will keep calling it a tree for now, because editing on my phone is hard], then you can calculate the desired subscription tree from the device tree, and then take a diff to determine which subscriptions to add or remove.

I'm assuming that there is a handshake for subscribing that goes 
```
subscribe ->
<- subscribed
```
or something. If that is the case then your subscription tree wants to have leaves like SubscriptionRequested (ts), Subscribed, Unsubscription requested, [Unsubscribed]

Where Unsubscribed might be represented by removing the node from the tree.

This would allow better separation of concerns, and simplify event handling. It would also make the topic subscription state machine explicit and observable when debugging is needed.

Reply:
I guess there is something like this handshake happening, but we are currently not looking at that level; we just assume that if we send a subscription request then the MQTT client library ensures that the server gets it and responds to it. Keeping track of that state here seems like making unnecessary work for ourselves which should already be handled at a lower layer.